WIP - Add Payment History dialog to Payments Tab (#2994)

willy-b opened the pull request:
- [x] Used Github [auto-closing keywords](https://help.github.com/articles/closing-issues-via-commit-messages/) in the commit message.

This pull request adds a Payment History dialog to the Payments Tab. 
It is a work in progress: The `Receipt Link` column links do not work yet (working on generating the receipt PDFs now). But the rest of the content in the Payment History dialog should be correct.

Detailed discussions and @bradleyrichter's original mockup can be found in issue #2994 (see mockup at https://github.com/brave/browser-laptop/issues/2994#issuecomment-241312583).

diracdeltas commented on app/ledger.js:
```diff
@@ -980,7 +981,10 @@ var getPaymentInfo = () => {
-        if ((body.rates) && (body.rates[currency])) info.btc = (amount / body.rates[currency]).toFixed(8)
+        if ((body.rates) && (body.rates[currency])) {
+          info.btcPrice = body.rates[currency]
+          info.btc = (amount / body.rates[currency]).toFixed(8)
```
need @mrose17 to review this part

diracdeltas replied:
i guess this is obsolete now.

mrose17 replied:
agreed. no longer needed.

diracdeltas commented on js/about/preferences.js:
```diff
@@ -530,6 +617,16 @@ class PaymentsTab extends ImmutableComponent {
+    return <div className='nextPaymentSubmission'><span>Your next payment submission is {nextReconcileDate}.</span></div>
+  }
+
   btcToCurrencyString (btc) {
```
you can simplify this using `btcPrice`

mrose17 replied:
actually `btcPrice` is now OBE based on what's in `master`

diracdeltas replied:
@mrose17 what do you mean? btcPrice was needed to convert `balance`/`unconfirmed` from BTC to local currency amount. if this has changed, please document modified ledger info properties in docs/state.md.

mrose17 replied:
+1

diracdeltas replied:
nvm comment above, @willy-b you can leave `btcToCurrencyString` unchanged

diracdeltas commented on js/about/preferences.js:
```diff
@@ -492,6 +573,12 @@ class PaymentsTab extends ImmutableComponent {
   }
 
+  get paymentHistoryButton () {
+    const buttonText = 'View Payment History'
```
please add this to l10n

diracdeltas commented on js/about/preferences.js:
```diff
@@ -576,6 +672,7 @@ class PaymentsTab extends ImmutableComponent {
                 {this.btcToCurrencyString(this.props.ledgerData.get('balance'))}
                 </span>
                 {this.walletButton}
+                {this.paymentHistoryButton}
```
this button should not show up if there is empty payment history (ex: before wallet has been created)

willy-b replied:
fixed in: https://github.com/brave/browser-laptop/pull/3391/commits/6c8274d4abb47a2063d4d24c55d6369083f7a89f#diff-e3eeb751016b2ce9f8278efce585a461R581

diracdeltas commented on js/about/preferences.js:
```diff
@@ -530,6 +617,16 @@ class PaymentsTab extends ImmutableComponent {
+  get paymentHistoryFooter () {
+    let ledgerData = this.props.ledgerData
+    let nextReconcileDate = formattedDateFromTimestamp(ledgerData.get('reconcileStamp'))
+    return <div className='nextPaymentSubmission'><span>Your next payment submission is {nextReconcileDate}.</span></div>
```
this should show nothing if `reconcileStamp` is undefined

diracdeltas replied:
please add the text to l10n

diracdeltas commented on js/about/preferences.js:
```diff
@@ -1128,4 +1233,12 @@ class AboutPreferences extends React.Component {
+  var dateStr = date.toDateString()
+  var [month, day, year] = dateStr.split(' ').slice(1)
+  dateStr = month + ' ' + day + ', ' + year
+  return dateStr
```
this may be hard to internationalize; i would rather just return `date.toLocaleDateString()` here

mrose17 replied:
+1

diracdeltas commented on js/about/preferences.js:
```diff
@@ -1071,6 +1174,7 @@ class AboutPreferences extends React.Component {
     if (overlayName === 'addFunds' && isVisible === false) {
       // Hide the child overlay when the parent is closed
       stateDiff['bitcoinOverlayVisible'] = false
+      stateDiff['paymentHistoryOverlayVisible'] = false
```
i don't think this line is necessary. the payment history overlay is unrelated to the add funds overlay.

diracdeltas commented on less/about/preferences.less:
```diff
@@ -455,6 +455,46 @@ table.sortableTable {
+  text-align: left;
+}
+
+span.paymentHistoryButton.browserButton {
```
this should not overflow when the window is narrow:
![screen shot 2016-08-25 at 10 42 37 am](https://cloud.githubusercontent.com/assets/549654/17979679/ca1913b0-6ab0-11e6-8c1b-b3f6859a8616.png)

willy-b replied:
Nice catch!
BTW, this screenshot points out another issue: The currency for the balance is BTC when the wallet is empty, and always USD when non-empty (just confirmed locally).

diracdeltas replied:
@willy-b before the wallet is created, we have no idea what the user's currency is, so it defaults to BTC. it could just say 'empty' or something.

diracdeltas replied:
this now looks awkwardly aligned with the select box but that can be saved for a follow-up issue.
![screen shot 2016-08-26 at 4 12 24 pm](https://cloud.githubusercontent.com/assets/549654/18023047/5ee1e924-6ba9-11e6-8e4b-7d5e21d3641a.png)

bradleyrichter replied:
yes...

Add Funds should be "Add funds..." and use smaller text. Button should match height of budget menu. Said menu should move up to align with "Add funds..." button.

diracdeltas commented:
I don't think we will add this to 0.11.6 @bbondy but looking good so far. Thanks @willy-b

willy-b commented:
Thanks for the code review @diracdeltas! I'll get those points addressed this afternoon.

mrose17 commented:
@willy-b what's your thinking in terms of a timeframe?

willy-b commented:
@mrose17, I will finish addressing the points @diracdeltas raised in her review by 3:30PM.

Then I aim to post a proposal/breakdown for the Receipt PDF based on @bradleyrichter's gorgeous new Receipt PDF mockups (reference: https://github.com/brave/browser-laptop/issues/2994#issuecomment-242287471) by 4:30PM.

For one I am going to suggest we omit absolute time figures for pages and only include contributions actually sent. This is the data available in the `ledgerData.transactions` array and also what I would argue is important to display (the probability of contributing to a site is less important _after_ the fact than what was contributed).

We might want to do simple CSV  receipts in this PR and implement @bradleyrichter's really nice mockups in a separate PR depending on schedule.

diracdeltas commented:
> (the probability of contributing to a site is less important after the fact than what was contributed

agree, and i think simple CSV receipts are fine for now

diracdeltas commented on app/ledger.js:
```diff
@@ -695,6 +695,7 @@ var ledgerInfo = {
 
   // the desired contribution (the btc value approximates the amount/currency designation)
   btc: undefined,
+  btcPrice: undefined,
```
@willy-b not sure if this is obvious to you, but the BTC price should now be available as a property of `transactions` according to @mrose17. specifically `ledgerInfo.transactions[i].contribution.rates` is the BTC price in the local currency at the time of the transaction (i think).

willy-b replied:
Yes, @mrose17's updates in ledger-client v0.8.46 brought in by #3397 are really nice and obsolete the `btcPrice` field I had added :-)

diracdeltas commented on js/about/preferences.js:
```diff
@@ -317,6 +317,87 @@ class BitcoinDashboard extends ImmutableComponent {
+
+  get totalAmount () {
+    var fiatAmount = this.satoshis * this.ledgerData.get('btcPrice') / 100000000
+    return fiatAmount.toFixed(2)
```
I think this is available directly now, via `this.transaction.get('contribution').get('fiat').get('amount')`. see docs/state.md

diracdeltas commented on js/about/preferences.js:
```diff
@@ -317,6 +317,87 @@ class BitcoinDashboard extends ImmutableComponent {
+  }
+
+  get satoshis () {
+    return this.transaction.get('satoshis')
```
@mrose17 is this correct, or should it be `this.transaction.get('contribution').get('satoshis')` now?

willy-b replied:
Sorry @diracdeltas, hold on just a minute, I haven't pushed the merge of @mrose17's update which includes this

willy-b replied:
@diracdeltas, while @mrose17 updated the `docs/state.md` in https://github.com/brave/browser-laptop/commit/5040058845acf9dd9700424a8a85940dc28457f7 , he didn't update  `app/ledger.js` to whitelist the `contribution` field to be picked from `ledger-client` transaction result. Thus after #3397 the `contribution` field still wasn't making it to UI.
I am updating the comments and code in `ledger.js` right now. Let me know if this is available in a different commit I can cherry-pick, because enabling `transactions[n].contribution` probably shouldn't be part of this PR.

diracdeltas replied:
will defer to @mrose17 on that

mrose17 replied:
oops! i'll fix it in master momentarily... sorry!

mrose17 replied:
use `this.transaction.get('contribution').get('satoshis')`

mrose17 commented on js/about/preferences.js:
```diff
@@ -539,8 +647,7 @@ class PaymentsTab extends ImmutableComponent {
-    if (this.props.ledgerData.get('btc') &&
-        typeof this.props.ledgerData.get('amount') === 'number') {
+    if (this.props.ledgerData.get('btc') && typeof this.props.ledgerData.get('amount') === 'number') {
       const btcValue = this.props.ledgerData.get('btc') / this.props.ledgerData.get('amount')
```
i suspect that sometime post-beta we will need to use something like https://github.com/xsolla/currency-format in order to format properly...

willy-b commented:
Update: I merged latest master to get @mrose17's updates in ledger-client v0.8.46 brought in by #3397, specifically the awesome new `transactions[n].contribution` field. 
But now I'm seeing some issues (`ledger-client` seems to be running over and over in a loop, and I'm trying to trigger a reconcile ahead of schedule to get real data with new format) so I'm running a bit behind schedule. (Only pushed 3 of the changes requested by @diracdeltas so far.)

mrose17 commented:
@willy-b - sorry about that. do me a favor and to your `ledger-state.json` file, towards the top look for 

      "options": {
        "verboseP": false,

and change `false` to `true` and then let me know what it's doing... thanks!

willy-b commented:
@mrose17 no worries at all, thanks for the help. 

It just appears to be logging the same line `ledger client run` over and over again:
```
ledger client run: clientP=true delayTime=0
ledger client run: clientP=true delayTime=0
ledger client run: clientP=true delayTime=0
...
```

mrose17 commented:
hmmm... i'd expect to see some network calls in there (lines starting with "<<<" or ">>>")

willy-b commented:
Ah, I see what it is. 

I was clearing the ledger-*.json files, opening the browser, navigating to pages, creating wallet, funding, and then looking at logs. But one has to navigate to pages and restart browser in order to have page history in the Ledger. 

If you don't have page history in the Ledger but the wallet is funded, then ledger-client spins.

willy-b commented:
Nevermind, ledger-client got caught in a loop again even with page history. I'l get back to you when I figure it out :-).

mrose17 commented:
Suggestion

In `run` have it throw an error if `delayTime === 0` so we can see the stack

diracdeltas commented:
@willy-b i just merged @mrose17's PR to whitelist contribution, so this needs another merge/rebase. sorry

mrose17 commented on app/ledger.js:
```diff
@@ -995,7 +1007,9 @@ var getPaymentInfo = () => {
-        if ((body.rates) && (body.rates[currency])) info.btc = (amount / body.rates[currency]).toFixed(8)
+        if ((body.rates) && (body.rates[currency])) {
+          info.btc = (amount / body.rates[currency]).toFixed(8)
+        }
```
i don't quite understand this change. it looks like 1 line got split to 3, but no other changes... what are my eyes not seeing? thanks!

willy-b replied:
Whoops! I had added a line for `btcPrice` there, then removed it after you added the `transactions[n].contribution` field that made `btcPrice` unnecessary. But it looks like I didn't restore the old code style properly :-)

diracdeltas commented on js/about/preferences.js:
```diff
@@ -323,6 +323,91 @@ class BitcoinDashboard extends ImmutableComponent {
+  }
+
+  get numericDateStr () {
+    var date = new Date(this.timestamp)
```
why not just return `date.toLocaleDateString()` here?

willy-b replied:
So I did do that for the displayed date: https://github.com/brave/browser-laptop/pull/3391/commits/48b0171bb5ad37071d360358522f242ea132ce7d#diff-e3eeb751016b2ce9f8278efce585a461R1258

This is for a filename, so `toLocaleDateString` is a little awkward with the slashes and lack of zero-padding. But sure, I can use `toLocaleDateString` and then replace the '/' with '-'s and let the lack of zero-padding be.

diracdeltas replied:
gotcha, i missed the fact that this is for a filename. sorry

bradleyrichter commented:
@willy-b Heads up for you. We are discussing a change that would swap the percent column for dollars (or local currency) per site. We should have this ironed out soon.

willy-b commented:
Thanks for heads up @bradleyrichter!
That's no problem. If I have the %, I can just multiply the % by the total in local currency. We do have the local currency value at historical exchange thanks to @mrose17's recent ledger-client update.

My semi-blocker right now is that the `ballots` field seems to have disappeared from the `transactions` entry -- that was what I used to get the %.

diracdeltas commented on js/about/preferences.js:
```diff
@@ -323,6 +323,83 @@ class BitcoinDashboard extends ImmutableComponent {
+  }
+
+  get satoshis () {
+    return this.transaction.get('contribution').get('satoshis')
```
`contribution` and other child properties may not necessarily exist, so this would throw `cannot read property 'get' of undefined` errors. please use `this.transaction.getIn(['contribution', ...])`.

diracdeltas commented on js/about/preferences.js:
```diff
@@ -323,6 +323,83 @@ class BitcoinDashboard extends ImmutableComponent {
+  }
+
+  get currency () {
+    return this.transaction.get('contribution').get('fiat').get('currency')
```
same as above

diracdeltas commented on js/about/preferences.js:
```diff
@@ -323,6 +323,83 @@ class BitcoinDashboard extends ImmutableComponent {
+  }
+
+  get totalAmount () {
+    var fiatAmount = this.transaction.get('contribution').get('fiat').get('amount')
```
same as above

diracdeltas commented on js/about/preferences.js:
```diff
@@ -323,6 +323,83 @@ class BitcoinDashboard extends ImmutableComponent {
+  render () {
+    var date = this.formattedDate
+    var totalAmountStr = this.totalAmount + ' ' + this.currency
+    var receiptFileName = 'brave_ledger' + this.numericDateStr + '.pdf'
```
this is minor, but we prefer to use JS template strings here. https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Template_literals

diracdeltas commented on js/about/preferences.js:
```diff
@@ -536,6 +627,27 @@ class PaymentsTab extends ImmutableComponent {
+  get paymentHistoryFooter () {
+    let ledgerData = this.props.ledgerData
+    if (!ledgerData.get('reconcileStamp')) {
+      return
```
return `null` here. https://facebook.github.io/react/docs/component-specs.html#render

diracdeltas commented on js/about/preferences.js:
```diff
@@ -323,6 +323,83 @@ class BitcoinDashboard extends ImmutableComponent {
+
+  get totalAmount () {
+    var fiatAmount = this.transaction.get('contribution').get('fiat').get('amount')
+    return fiatAmount.toFixed(2)
```
this should check that `fiatAmount` is a Number

diracdeltas commented on less/about/preferences.less:
```diff
@@ -455,6 +455,61 @@ table.sortableTable {
+    padding-bottom: 10px;
+  }
+
+  .dialog-body {
```
this needs `overflow-y: scroll;` or else this happens:
![screen shot 2016-08-26 at 4 27 33 pm](https://cloud.githubusercontent.com/assets/549654/18023141/3028338a-6baa-11e6-94c2-99345d57e5e9.png)

willy-b replied:
Your recommendation fixed this:  https://github.com/brave/browser-laptop/pull/3391/files/1eded36027dbe9bf21d1fdfeb8fdcec8a3ff9a61#r76502448

willy-b replied:
Your recommendation fixed this: https://github.com/brave/browser-laptop/pull/3391/commits/2a729d9aacc94127d7bddb5338449cd915235f8c#diff-4b50ff14dadc9756719a577a72f520d6R500

diracdeltas commented on js/components/modalOverlay.js:
```diff
@@ -18,10 +18,19 @@ class ModalOverlay extends ImmutableComponent {
-      {this.props.content}
+    let customTitleClassesStr = (this.props.customTitleClasses ? this.props.customTitleClasses : '')
+
+    return <div className={'dialog ' + customTitleClassesStr}>
```
something in this ModalOverlay change is breaking the `Buy with Coinbase` modal overlay:
![screen shot 2016-08-26 at 4 29 46 pm](https://cloud.githubusercontent.com/assets/549654/18023153/6bd7051e-6baa-11e6-8fc7-4ced55e1fdca.png)

willy-b replied:
Wow, glad you caught this :-).
Fixed by https://github.com/brave/browser-laptop/pull/3391/commits/2a729d9aacc94127d7bddb5338449cd915235f8c#diff-4b50ff14dadc9756719a577a72f520d6R513

diracdeltas commented:
@willy-b does the semi-blocker relating to the disappearance of `ballots` only affect the transaction CSV/PDF files? I am fine with merging this for now once the comments above are addressed, and the CSV/PDF part can be done in a follow-up PR.

Thanks for the work; please run `git rebase -i` to squash commits when you're done.

willy-b commented:
Thanks for all the testing and code reviewing @diracdeltas. 

Yes, the `ballots` issue only affects the transaction-level CSV/PDF. If everyone is OK with it, I think it's a great idea to move the 'Receipt Link' to a separate PR.

bradleyrichter commented:
@willy-b False alarm on the change to currency. We are staying with percentages for now.

bbondy commented:
Is this ready to merge? I won't include it in beta5 yet because I'm not sure.

mrose17 commented:
@willy-b - what's your thinking on @bbondy's question?

willy-b commented:
Hey! If we are OK with doing the 'Receipt Link' in a separate PR then I
just need to disable that column and squash the commits for this to undergo
a final review for merging.
On Aug 26, 2016 10:32 PM, "Marshall T. Rose" <notifications@github.com>
wrote:

> @willy-b <https://github.com/willy-b> - what's your thinking on @bbondy
> <https://github.com/bbondy>'s question?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/brave/browser-laptop/pull/3391#issuecomment-242897666>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AGpaA-_nvMZsQjzogvuCTwD4frW74QBSks5qj8vqgaJpZM4Jskfu>
> .
>

mrose17 commented:
@willy-b - i like that plan... please proceed

cc: @bbondy

diracdeltas commented:
@willy-b agree; that sounds like a good plan. you can just hide the receipt link column for now.

willy-b commented:
Sorry for the delay, coming right up!

On Fri, Aug 26, 2016 at 11:41 PM, yan zhu (@bcrypt) <
notifications@github.com> wrote:

> @willy-b <https://github.com/willy-b> agree; that sounds like a good
> plan. you can just hide the receipt link column for now.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/brave/browser-laptop/pull/3391#issuecomment-242900147>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AGpaA0gMsGvtxGxEXcl_F2jeOsegjwyVks5qj9wogaJpZM4Jskfu>
> .
>

willy-b commented:
The 'Receipt Link' has been disabled and this has been rebased down to two commits ready for review in PR  #3473. 

Thanks again @diracdeltas and @mrose17 for the code reviews / testing, and for helping me come up to speed on your project as I worked on this :-).
